### PR TITLE
Add the `__repr__` method for Instructions

### LINKED ISSUE #6920: Define `__str__` for `Instruction` objects
### What is the expected enhancement?

Printing a `Instruction` to console shows the default pointer information because `__str__` is not defined. It would be useful to show at least the minimum information of the instruction itself, i.e. `self.name`, `self.num_qubits` and `self.num_clbits`.

### PR DESCRIPTION
**Pull request by TheGupta2012**
### Summary
Fixes #6920 by adding an `__repr__` method for Instruction class

### Details and comments
- This PR adds a basic `__repr__` method for the `Instruction` class
- Information provided is the Instruction name, number of qubits, number of classical bits and the parameter map of Instruction.
- Example :

```
from qiskit import * 
qc = QuantumCircuit(3,name="my-instruction-circuit")
qc.cx([0,1],[1,2])
qc.measure_all()
ins = qc.to_instruction()
```  

image

### CODE REVIEW

**Review by kdk — COMMENTED** [review]
Thanks @TheGupta2012 , this will be really helpful! It might more sense to implement this as a `__repr__` method (which will then automatically handle `__str__`) following from https://docs.python.org/3/reference/datamodel.html#object.__repr__  (so the expected output might look something like `Instruction(name='rx', num_qubits=1, num_clbits=0, params=[3.14159])`. Can you also add a test?

**Comment by TheGupta2012** [comment]
Even I was thinking that as the `__repr__` also works inline when working in a jupyter notebook.
Also, I am a bit unaware of how tests are written. Could you describe a bit what the structure of a test looks like? Thank you!

**Comment by kdk** [comment]
As an example, see: https://github.com/Qiskit/qiskit-terra/blob/main/CONTRIBUTING.md#test and https://github.com/Qiskit/qiskit-terra/blob/a34b2ab002c0e0935d02691026ae676ee0adf90b/test/python/circuit/test_circuit_data.py#L154 .

**Comment by TheGupta2012** [comment]
I have added a `__repr__` method as specified above.
The structure followed is exactly the same as `Instruction(name='abc' num_qubits=1, num_clbits=0, params=[3.14159])`

File: qiskit/circuit/instruction.py [hunk 1]
```diff
@@ -144,6 +144,16 @@ def __eq__(self, other):
 
         return True
 
+    def __repr__(self) -> str:
+        """Generates a representation of the Intruction object instance
+        Returns:
+            str: A representation of the Instruction instance with the name,
+                 number of qubits, classical bits and params( if any )
+        """
+        return "Instruction(name = '{}', num_qubits = {}, num_clbits = {}, params = {})".format(
```
Comment (kdk) [inline qiskit/circuit/instruction.py:153]:
```suggestion
        return "Instruction(name='{}', num_qubits={}, num_clbits={}, params={})".format(
```
This looks good, but I think would benefit from being a little more compact. E.g.
```
Instruction(name='rx', num_qubits=1, num_clbits=0, params=[3.14159])
```
vs
```
Instruction(name = 'rx', num_qubits = 1, num_clbits = 0, params = [3.14159])
```

**Review by kdk — COMMENTED** [review]
Thanks for the update @TheGupta2012 , I think this looks good. One small formatting comment.

File: qiskit/circuit/instruction.py (hunk 1 above)
Comment (TheGupta2012) [reply qiskit/circuit/instruction.py:153]: Okay looks good, will update it today!

**Review by kdk — APPROVED** [review]
Thanks for the updates @TheGupta2012 , this looks good!

**Pull request merged**